### PR TITLE
Benchmark improvements

### PR DESCRIPTION
**Pull request by prashantv**
- Support running a no-library version of client and server (uses templated TChannel frames)
 - Add dummy relay benchmarks that just relay TCP

### CODE REVIEW

File: benchmark/tcp_bench_test.go [hunk 1]
```diff
@@ -46,17 +47,26 @@ func echoServer(tb testing.TB) net.Listener {
...
 	conn, err := net.Dial("tcp", dst)
 	require.NoError(b, err, "Failed to connect to destination")
+	defer conn.Close()
 
-	go io.Copy(ioutil.Discard, conn)
+	readerDone := make(chan struct{})
+	go func() {
+		defer close(readerDone)
+		_, err := io.CopyN(ioutil.Discard, conn, int64(totalExpected))
+		assert.NoError(b, err, "Expected more response bytes")
```
Comment (akshayjshah) [inline benchmark/tcp_bench_test.go]: The error message might be nicer if we captured the number of bytes successfully written ("Expected 1024 response bytes, got 0").

Comment (prashantv) [inline benchmark/tcp_bench_test.go]: 👍  done

File: testutils/counter.go [hunk 2]
```diff
@@ -66,14 +66,14 @@ func Decrementor(n int) Decrement {
 	}
 }
 
-// Batches returns a slice with n broken into batches of size at most batch.
-func Batches(n, batch int) []int {
-	numConstantBatches := n / batch
-	batches := make([]int, 0, numConstantBatches+1)
-	for i := 0; i < numConstantBatches; i++ {
-		batches = append(batches, batch)
+// Batch returns a slice with n broken into batches of size at most batch.
```
Comment (akshayjshah) [inline testutils/counter.go]: nit: the parameter is now called `batchSize`.

File: testutils/counter.go [hunk 3]
```diff
@@ -79,6 +79,20 @@ func Batch(n, batchSize int) []int {
...
+
+	buckets := make([]int, numBuckets)
+	for i := range buckets {
+		buckets[i] = perBucket
+		if i == 0 {
+			buckets[i] += n % numBuckets
+		}
+	}
+	return buckets
+}
```
Comment (akshayjshah) [inline testutils/counter.go]: :+1:

File: benchmark/req_bytes.go [hunk 4]
```diff
@@ -31,5 +29,10 @@ func getRequestBytes(n int) []byte {
 }
 
 func getRequestString(n int) string {
-	return base64.RawStdEncoding.EncodeToString(getRequestBytes(n))[:n]
+	chars := []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcedefghijklmnopqrstuvwxyz")
+	bs := make([]byte, n)
+	for i := range bs {
+		bs[i] = chars[i%len(chars)]
+	}
+	return string(bs)
```
Comment (akshayjshah) [inline benchmark/req_bytes.go]: Can we leave a TODO to replace this with base64 encoding once we drop Go 1.4 support?

File: testutils/counter.go (hunk 2 above)
Comment (prashantv) [inline testutils/counter.go]: done

File: benchmark/req_bytes.go (hunk 4 above)
Comment (prashantv) [inline benchmark/req_bytes.go]: added

**Comment by akshayjshah** [comment]
lgtm...this is super awesome.

**Pull request merged**